js/services.js

```diff
@@ -0,0 +1,166 @@
+/* global ServiceConfig, Promise, Messenger */
+
+'use strict';
+
+(function(exports) {
+  /**
+   * This module is responsible to create and run services.
+   */
+  var Services = function() {};
+  Services.prototype = {
+
+    _dispatchReady: function() {
+      // With all important event handlers in place, we can now notify
+      // Gecko that we're ready for certain system services to send us
+      // messages (e.g. the radio).
+      // Note that shell.js starts listen for the mozContentEvent event at
+      // mozbrowserloadstart, which sometimes does not happen till
+      // window.onload.
+      var evt = new CustomEvent('mozContentEvent', {
+        bubbles: true, cancelable: false,
+        detail: { type: 'system-message-listener-ready'}
+      });
+      window.dispatchEvent(evt);
+    },
+
+    handleBrowserLoaded: function(instance) {
+      if (instance.config.type === 'window' ||
+          instance.config.type === 'app') {
+        instance.iframe.hidden = false;
+        instance.iframe.setVisible(true);
+      }
+
+      if (instance.config.type === 'app') {
+        // We cannot create messenger for an app since it is at remote
+        // process.
+        // move focus to foreground window
+        document.activeElement.blur();
+        instance.iframe.focus();
+        // process next one
+        this._processCreateQueue();
+        return;
+      }
+      // create messenger
+      instance.messenger = new Messenger();
+      instance.messenger.start(instance.iframe);
+      instance.messenger.onPingBack = function() {
+        this.send({
+          'log': 'service [' + instance.config.manifestURL + '] pinged'
+        });
+        if (instance.config.type === 'window') {
+          // move focus to pinged window
+          document.activeElement.blur();
+          instance.iframe.focus();
+        }
+        // process next one
+        this._processCreateQueue();
+      }.bind(this);
+    },
+
+    handleBrowserError: function(instance) {
+      if (instance.config.type === 'window' ||
+          instance.config.type === 'app') {
+        // move focus to top most window
+        document.activeElement.blur();
+        var iframe = this._windowContainer.querySelector(
+                                                       'iframe:last-child');
+        if (iframe) {
+          iframe.focus();
+        }
+      }
+      // remove the old one and recreate it.
+      instance.messenger.stop();
+
+      instance.iframe.parentElement.removeChild(instance.iframe);
+      if (instance.config.autoRestart) {
+        this._createService(instance.config);
+      }
+    },
+
+    handleEvent: function(e) {
+      var instance = this._serviceMap[e.target.getAttribute('mozapp')];
```

Comment:
wish oneday we need to wrap service mapping to another method :dancer: 
Its totally fine now.